wasm_builder/src/lib.rs

```diff
@@ -280,16 +320,46 @@ impl Output {
     /// Fails if internal tool fails to optimize wasm binary.
     #[allow(clippy::unnecessary_wraps)]
     pub fn optimize(self) -> Result<Self> {
-        let optimizer = wasm_opt::OptimizationOptions::new_optimize_for_size();
+        let optimized_file = PathBuf::from(format!(
+            "{parent}{separator}{file}_optimized.{ext}",
+            parent = self
+                .wasm_file
+                .parent()
+                .map_or_else(String::default, |p| p.display().to_string()),
+            separator = std::path::MAIN_SEPARATOR,
+            file = self
+                .wasm_file
+                .file_stem()
+                .map_or_else(|| "output".into(), OsStr::to_string_lossy),
+            ext = self
+                .wasm_file
+                .extension()
+                .map_or_else(|| "wasm".into(), OsStr::to_string_lossy),
+        ));
+
+        if let Some(previous_hash) = &self.previous_hash {
+            let current_hash =
+                sha256::try_digest(self.wasm_file.as_path()).wrap_err_with(|| {
+                    format!(
+                        "Failed to compute sha256 digest of wasm file: {}",
+                        self.wasm_file.display()
+                    )
+                })?;
+
+            if optimized_file.exists() && &current_hash == previous_hash {
+                return Ok(Self {
+                    wasm_file: optimized_file,
+                    previous_hash: self.previous_hash,
+                });
+            }
+        }
 
-        let optimized_file = tempfile::NamedTempFile::new_in(self.out_dir.path())?
-            .path()
-            .to_owned();
+        let optimizer = wasm_opt::OptimizationOptions::new_optimize_for_size();
         optimizer.run(self.wasm_file, optimized_file.as_path())?;
 
         Ok(Self {
             wasm_file: optimized_file,
-            out_dir: self.out_dir,
+            previous_hash: self.previous_hash,
```

Comment:
I don't understand why we didn't return a new hash as `previous_hash` here. According to the condition above, the `previous_hash` equals `None` or an outdated hash, it isn't?
```rust
Ok(Self {
    wasm_file: optimized_file,
    previous_hash: new_hash, // or reuse `current_hash`
})
```